clusterloader/docs/vision.md

```diff
@@ -0,0 +1,322 @@
+tests to that framework.
+
+We would like to get back to that and build fully functional Cluster Loader and
+use it as a framework to run all our scalability tests. This doc is describing
+high-level vision of how this will work.
+
+
+## Vision
+
+At the high level, a single test will consist of a number of phases (if we have
+really good usecase we may consider allowing the to create a DAG, but for now
+I don’t have rationale for it). In each phase we will be creating/updating/deleting
+a number of different objects.
+Additionally, we will introduce a set of predefined operations (that user will be
+able to use as phases). They will allow users to monitor/measure performance impact
+of user-defined phases.
+The following subsections will describe this in a bit more details.
+
+
+### Tuning Set
+
+Since we would like to be fully load even very big clusters, we need to be able to
+create a number of “similar” objects. A “Tuning Set” concept will allow us to
+spread those operations across some time.
+We define Tuning Set as following:
+
+```
+struct TuningSet {
+	Name string
+	// Exactly one of the following should be set.
+	QpsLoad        *QpsLoad
+	RandomizedLoad *RandomizedLoad
+	SteppedLoad    *SteppedLoad
+}
+
+// QpsLoad defines a uniform load with a given QPS.
+struct QpsLoad {
+	InitialDelay time.Duration
+	Qps          float
+}
+
+// RandomizedLoad defines a load that is spread randomly
+// across a given total time.
+struct RandomizedLoad {
+	InitialDelay time.Duration
+	AverageQps   float
+}
+
+// SteppedLoad defines a load that generates a burst of
+// a given size every X secods.
+struct SteppedLoad {
+	BurstSize    int32
+	InitialDelay time.Duration
+	StepDelay    time.Duration
+}
+```
+
+More policies can be introduced in the future.
+
+
+### Multiple copies of the same workload
+
+To fill-in large clusters, we would need spread objects across different namespaces.
+In many cases, it will be enough for users for many namespaces to contain the same
+objects (or to be more specific: objects created from the same templates). Obviously,
+we want the config for the test to be as small as possible.
+As a result, we will introduce the following rules:
+- In the top-level test definition, we will define the number of namespaces that will
+  be used by a test.
+- All the namespaces will have names of a form “namespace-<number>” for number
+  in range 1..N (where N is the number of namespaces in a test)
+
+To make it usable, we introduce the following type:
+
+```
+struct Range {
+	Min int32
+	Max int32
+}
+```
+
+In the context of namespaces, such object defines all namespaces "namespace-\<i\>" for
+`i` in the range [Min, Max].
+
+
+### Phase
+Since phase will consist of a number of create, update and delete operations (potentially
+on many different object types). So at the high level description of phase will be
+defined as:
+
+```
+struct Phase {
+	Declarations []PhaseDeclare
```

Comment:
The same here.. You can simply have `Step = { *PhaseDeclare or *Measurement }` instead of one more struct and array.

Reply:
That isn't true - this would reduce what can be expressed.
Having here []PhaseDeclare allows you to e.g. in parallel create one set of object, create the second set of objects and delete the third of objects.
So basically, PhaseDeclare`s within a single phase are being executed in parallel.

Reply:
That said, we can probably get rid of `Phase` here and change Step to be either []PhaseDeclare or Measurement.

I will change that and make explicit in the doc.

Reply:
> I will change that and make explicit in the doc.

Yes, thanks.

And wrt your parallelization argument, let me think a bit more if this is the way we want to formulate it.

Reply:
Actually, I simplified it a bit. Now we only have a concept of "Step", "Phase" and "Measurement" and I got rid of "PhaseDeclare" at all.

PTAL